golang1.17/lib/launcher.go

```diff
@@ -227,6 +230,37 @@ func buildArguments(ctx context.Context, f reflect.Value, in []byte) ([]reflect.
 	return []reflect.Value{reflect.ValueOf(val).Elem()}, nil
 }
 
+// buildContext takes the context that is being used to invoke the user's function and adds the "context" data to it.
+//
+// Go's convention for contexts is to define a struct in the package you want to build contexts from and use a constant
+// instance of that struct as the key and define an exported function that other packages can use to retrieve a struct
+// with the data you want to convey from the context.
+// See https://github.com/aws/aws-lambda-go/blob/main/lambdacontext/context.go for an example of this.
+// If you need more than one key per package, you can define a struct type that you can put something like a string in
+// so that you can differentiate constant instances of the structs.
+// See https://github.com/golang/go/blob/master/src/net/http/server.go for an example of this.
+//
+// At DO, we don't yet offer a library our users can import as they write their functions, so we can't follow this
+// pattern right now. Therefore, the naming convention we use for our context keys doesn't matter. In the future, we can
+// change this by creating a library that our customers can use, at which point we will no longer have multiple context
+// keys and the single remaining context key will no longer be a string. Therefore, while we do need to communicate
+// these string keys to our customers for now, the naming convention we use for the strings doesn't matter. We've chosen
+// this-case arbitrarily.
+func buildContext(ctx context.Context) context.Context {
```

Comment:
Should we subsume adding the deadline to the context into this function as well?

Reply:
Agree. I moved all of the context building into one place. To keep things as idiomatic as possible for Go, I have the context built in the `execute` function. I feel like it's idiomatic to construct a context, with a deadline if you need it, before you use the context in a function call you might need to cancel.